source/server/listener_manager_impl.cc

```diff
@@ -843,14 +867,29 @@ std::unique_ptr<Network::FilterChain> ListenerFilterChainFactoryBuilder::buildFi
   ProtobufTypes::MessagePtr message = Config::Utility::translateToFactoryConfig(
       transport_socket, parent_.messageValidationVisitor(), config_factory);
 
-  std::vector<std::string> server_names(filter_chain.filter_chain_match().server_names().begin(),
-                                        filter_chain.filter_chain_match().server_names().end());
-
-  return std::make_unique<FilterChainImpl>(
+  return std::make_unique<TcpFilterChainImpl>(
       config_factory.createTransportSocketFactory(*message, factory_context_,
                                                   std::move(server_names)),
       parent_.parent_.factory_.createNetworkFilterFactoryList(filter_chain.filters(), parent_));
 }
 
+QuicListenerFilterChainFactoryBuilder::QuicListenerFilterChainFactoryBuilder(
+    ListenerImpl& listener,
+    Server::Configuration::TransportSocketFactoryContextImpl& factory_context)
+    : ListenerFilterChainFactoryBuilder(listener, factory_context) {}
+
+std::unique_ptr<Network::FilterChain> QuicListenerFilterChainFactoryBuilder::buildFilterChain(
+    const ::envoy::api::v2::listener::FilterChain& filter_chain) const {
+  // Initialize filter chain for QUIC.
+  if (!filter_chain.has_tls_context()) {
+    return nullptr;
```

Comment:
Should this be a config error?

Reply:
We should check this during config load and fail the config, it should be an assert at this point IMO.

Reply:
With removal of QuicFilterChain, this builder can be unified with tcp one too. This check no longer apply.